docs/conversion/sakai_20_oracle_conversion.sql

```diff
@@ -13,6 +13,12 @@ ALTER TABLE POLL_OPTION ADD OPTION_ORDER NUMBER(10,0);
 
 -- END SAK-41391
 
+-- SAK-47205
+-- This will target orphaned poll_option rows that the Backfill poll option order quartz job may miss.
+-- The Backfill poll option order job should be run after applying this conversion
+DELETE from poll_option where option_id in (select * from (select a.option_id from poll_option as a left outer join poll_poll as b on a.option_poll_id = b.poll_id where b.poll_id is null) as t);
```

Comment:
```suggestion
DELETE FROM poll_option
WHERE option_id IN (
    SELECT * FROM (
        SELECT a.option_id
        FROM poll_option a
        LEFT OUTER JOIN poll_poll b
        ON a.option_poll_id = b.poll_id
        WHERE b.poll_id IS NULL
        )
    t);
```

Reply:
I'm still really confused how this is even possible. How can `poll_poll.poll_id` be null? It's the primary key of the table! When @austin48 described this issue to me, I assumed that the `poll_poll.site_id` pointed to a site that no longer existed, and thus the job would not pick this poll up because there was no site attached to it (when the job was iterating by site).

Reply:
poll_id is not null in either of the "real" tables, poll_poll nor poll_option.  In MySQL, the "left outer join"'s temp. table will show the poll_id as null on the "right side" of the join if it didn't find any matches.  Maybe it would have been clearer if I used _another_ nested select instead of the outer join?

Also, it would be interesting to see if anyone else finds matches in _their_ databases by substituting the delete for a select.  Especially if their Sakai instance has been upgraded through the years and has auto.ddl=false.

Reply:
Okay now I'm even more confused. If poll_id is not null in the "real" tables, what are we really targetting here? Is it that the options belong to now polls? Or the polls belong to no sites? It's a little unclear what the circumstances you're trying to deal with are.

I can tell you that we do exactly that: we upgrade our databases via scripts, and we leave auto.ddl=false at all times. We do not have any orphaned polls, nor orphaned poll options in any of our databases.

I don't understand the necessity of the left outer join. If the issue is that the options belong to no polls, simply query for poll_poll.poll_id == null. If the issue is that the polls belong to sites that no longer exist, simply query for polls where the corresponding site_id in sakai_site is null.

Am I missing something here?

Reply:
> the middle SELECT * is likely unnecessary

in MySQL, that part was necessary because when I was still setting to 0 instead of deleting, mysql doesn't let you update the same table that is in the nested select.  The SELECT * is a work around.  I believe it creates a temp. table that then let the update happen.  I assumed it would be the same for a delete and left that part in.

Reply:
our poll_option table has some records where it's option_poll_id no longer exists in poll_poll

the issue is that after running the backfill job (including my java patch in this jira), that job will get a list of all poll_poll records, then for each, it will re-do the option_order (e.g. fix nulls).  But since we have poll_option orphans, those options will not be modified (their parents weren't in the list of poll_poll rows) and their order will still be null.  Then the Sakai 21 conv. will fail because it not longer accepts nulls.  (Maybe on the java side, you could get a list of _all_ poll_options grouped by option_poll_id instead of all the poll_poll rows, but then you'd need to add a new function to get that query?)

to find orphans you can the nested selects in my original code or run:
```
SELECT option_id, option_poll_Id FROM poll_option
WHERE option_id IN (
    SELECT a.option_id
    FROM poll_option a
    WHERE a.option_poll_id NOT IN (
        SELECT b.poll_id
        FROM poll_poll as b
    )
);
```

Reply:
DELETE from POLL_OPTION where OPTION_POLL_ID NOT IN (SELECT POLL_ID FROM POLL_POLL)